File: aoc2022/day01.jl
test = """
    1000
    2000
    3000

    4000

    5000
    6000

    7000
    8000
    9000

    10000
"""

function parse_input(input::AbstractString)
    lines = split(input, '\n') .|> strip
    breaks = findall(isempty, lines)
    groups = map([0; breaks], [breaks; length(lines) + 1]) do from, to
        return parse.(Int, lines[from + 1:to - 1])
    end
    return filter(!isempty, groups)
end

q1(groups) = maximum(sum.(groups))
q2(groups) = sum(partialsort!(sum.(groups), 1:3; rev = true))

let v = parse_input(test)
    @assert q1(v) == 24000
    @assert q2(v) == 45000
end

v = parse_input(read("day01.in", String))
println("Q1: ", q1(v))
println("Q2: ", q2(v))

File: aoc2022/day01.rs
use std::io::{self, BufRead};

// fn main() {
//     // Read the input from stdin.
//     let stdin = io::stdin();
//     let lines: Vec<String> = stdin.lock().lines().map(|line| line.unwrap()).collect();

//     // Parse the input and compute the result.
//     let result:u64 = lines
//         .split(|line| line.is_empty())
//         .map(|group| group.iter().map(|line| line.parse::<u64>().unwrap()).sum())
//         .max()
//         .unwrap();

//     // Print the result to stdout.
//     println!("{}", result);
// }

fn main() {
    // Read the input from stdin.
    let stdin = io::stdin();
    let lines: Vec<String> = stdin.lock().lines().map(|line| line.unwrap()).collect();

    // Parse the input and compute the result.
    let mut result: Vec<u64> = Vec::new();

    for group in lines.split(|line| line.is_empty()) {
        let total: u64 = group.iter().map(|line| line.parse::<u64>().unwrap()).sum();
        result.push(total);
    }

    result.sort_unstable_by(|a, b| b.cmp(a));

    let result:u64 = result[0..3].iter().sum();

    // Print the result to stdout.
    println!("{}", result);
}
